pkg/validatingadmissionpolicy/validate.go

```diff
@@ -152,30 +159,136 @@ func Validate(
 			v1binding := ConvertValidatingAdmissionPolicyBinding(binding)
 			isMatch, err := matcher.BindingMatches(a, o, validating.NewValidatingAdmissionPolicyBindingAccessor(&v1binding))
 			if err != nil {
-				return engineResponse, err
+				return nil, err
 			}
 			if !isMatch {
 				continue
 			}
+			if binding.Spec.ParamRef != nil {
+				params, err := CollectParams(context.TODO(), adapters.Client(client), policy.Spec.ParamKind, binding.Spec.ParamRef, resource.GetNamespace())
+				if err != nil {
+					return nil, err
+				}
+				logger.V(3).Info("validate resource %s against policy %s with binding %s", resPath, policy.GetName(), binding.GetName())
+				for _, p := range params {
+					engineResponse, err := validateResource(policy, &bindings[i], resource, *namespace, a, p)
+					if err != nil {
+						continue // is this correct ?
+					}
+					ers = append(ers, engineResponse)
+				}
+				return ers, nil
+			}
 
-			logger.V(3).Info("validate resource %s against policy %s with binding %s", resPath, policy.GetName(), binding.GetName())
-			return validateResource(policy, &bindings[i], resource, *namespace, a)
+			engineResponse, err := validateResource(policy, &bindings[i], resource, *namespace, a, nil)
+			if err != nil {
+				continue // is this correct ?
+			}
+			ers = append(ers, engineResponse)
+			return ers, nil
 		}
 	} else {
 		for i, binding := range bindings {
-			isMatch, err := matches(a, namespaceSelectorMap, *binding.Spec.MatchResources)
+			// check if its a global binding
+			if binding.Spec.MatchResources != nil {
+				isMatch, err := matches(a, namespaceSelectorMap, *binding.Spec.MatchResources)
+				if err != nil {
+					return nil, err
+				}
+				if !isMatch {
+					continue
+				}
+			}
+
+			if binding.Spec.ParamRef != nil {
+				params, err := CollectParams(context.TODO(), adapters.Client(client), policy.Spec.ParamKind, binding.Spec.ParamRef, resource.GetNamespace())
+				if err != nil {
+					return nil, err
+				}
```

Comment:
In the case of offline mode, there is no need to call `CollectParams()`, we should fetch the parameter resources from the files passed in the flag `--resource` or `--parameterResource`.

First, you have to add parameter resources in the `PolicyData`.
https://github.com/kyverno/kyverno/blob/deab3fd467145109ac96b8666f8a56e2a466cbd5/pkg/validatingadmissionpolicy/api.go#L14-L19

And, implement the function `AddParameterResources()` similar to what we have for bindings:
https://github.com/kyverno/kyverno/blob/deab3fd467145109ac96b8666f8a56e2a466cbd5/pkg/validatingadmissionpolicy/api.go#L20-L23

Now, you have the parameter resources in the `PolicyData` and you can use it in case of the offline mode (i.e. `client` is nil). There is no need for the parameter `isCluster`. That's why we don't have to use fake clients in the 1st place.

Here is how we add bindings to the `PolicyData`:
https://github.com/kyverno/kyverno/blob/e3a86bc286bdae91d41146b0bd378229152063ab/cmd/cli/kubectl-kyverno/processor/vap_processor.go#L25-L29

When we add bindings, we can get the policy paramKind and the binding paramRef and use this information to fetch the resource from the file path and then you can add it in the `PolicyData`.

Reply:
It's a perfectly rational suggestion, i will go with it if we see that the need to unify the logic isn't that pressing.
As i mentioned in the other comment i think it makes the code simpler to read. Since you only have to trace one line of logic (the interaction with some k8s client)

Reply:
We're creating a fake client to fetch parameter resources, even though we can fetch it from the input files (offline mode).

I mean in case of using fake clients, we will have to call `client.GetResource()`, and `client.ListResource()` to fetch the parameter resource although we can fetch it from the input files.

Reply:
[This](https://github.com/kyverno/kyverno/blob/deab3fd467145109ac96b8666f8a56e2a466cbd5/pkg/validatingadmissionpolicy/validate.go#L231-L292) function is used offline and I believe it's unnecessarily expensive because the parameter resource could be retrieved directly from the input, avoiding the overhead of creating a fake client and making potentially costly calls to retrieve it.